gadopt/stokes_integrators.py

```diff
@@ -195,4 +211,58 @@ def setup_solver(self):
     def solve(self):
         if not self._solver_setup:
             self.setup_solver()
+
         self.solver.solve()
+
+
+class FreeSurfaceStokesSolver(StokesSolver):
+    name = 'FreeSurfaceStokesSolver'
+
+    def __init__(self, z, T, approximation, free_surface_dt, free_surface_id, bcs=None, mu=1,
+                 quad_degree=6, cartesian=True, solver_parameters=None, closed=True,
+                 rotational=False, J=None, constant_jacobian=False, **kwargs):
+
+        self.free_surface_id = free_surface_id
+
+        super().__init__(z, T, approximation, bcs=bcs, mu=mu, quad_degree=quad_degree,
+                         cartesian=cartesian, solver_parameters=solver_parameters, closed=closed,
+                         rotational=rotational, J=J, constant_jacobian=constant_jacobian, **kwargs)
+
+        # Add free surface time derivative term
+        self.F += self.equations[2].mass_term(self.test[2], (self.eta-self.eta_old)/free_surface_dt)
+
+    def setup_equations(self):
+        # Initialise Stokes equations with a free surface, this method will be called instead of the
+        # StokesSolver.setup_equations() in StokesSolver.__init__()
+
+        # First setup Stokes equations (without a free surface)
+        super().setup_equations()
+
+        # Define free surface variables for timestepping
+        assert len(self.stokes_vars) == 3
+        self.solution_old = fd.Function(self.solution)
+        self.eta = self.stokes_vars[2]
+        self.eta_old = fd.split(self.solution_old)[2]
```

Comment:
Do we actually need `solution_old`, or can we just use `eta_old`?

Reply:
I think when I tried this I couldn't get the assign to work from eta (part of the mixed space) to eta_old though I might have done something stupid. I agree I think it would maybe be neater with just eta for now as we don't need the old velocity or pressure.

Reply:
What if you have `self.eta_old = fd.Function(self.eta)`?

Reply:
I think I tried that and it didn't work because of the mixed spaces... But maybe I should try again